File: crates/normalize-languages/tests/fixtures/matlab/sample.m
% Sample MATLAB file

classdef Shape < handle
    % A base class for geometric shapes
    properties
        Color = 'blue'
    end

    methods
        function obj = Shape(color)
            if nargin > 0
                obj.Color = color;
            end
        end

        function area = computeArea(obj)
            area = 0;
        end

        function display(obj)
            fprintf('Shape: color=%s, area=%g\n', obj.Color, obj.computeArea());
        end
    end
end

function result = factorial(n)
% Compute factorial of n
    if n <= 1
        result = 1;
    else
        result = n * factorial(n - 1);
    end
end

function total = sumArray(arr)
% Return sum of all elements in arr
    total = 0;
    for k = 1:length(arr)
        total = total + arr(k);
    end
end

function filtered = filterPositive(arr)
% Return only positive elements of arr
    filtered = [];
    for k = 1:length(arr)
        if arr(k) > 0
            filtered(end + 1) = arr(k); %#ok<AGROW>
        end
    end
end

function result = classifyNumber(n)
% Classify a number as negative, zero, or positive
    switch sign(n)
        case -1
            result = 'negative';
        case 0
            result = 'zero';
        otherwise
            result = 'positive';
    end
end

File: crates/normalize-languages/tests/fixtures/perl/sample.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw(sum max min);
use POSIX qw(floor ceil);

package Calculator;

sub new {
    my ($class, %args) = @_;
    return bless {
        precision => $args{precision} // 2,
    }, $class;
}

sub add {
    my ($self, $a, $b) = @_;
    return $a + $b;
}

sub multiply {
    my ($self, $a, $b) = @_;
    return $a * $b;
}

package main;

sub classify {
    my ($n) = @_;
    if ($n < 0) {
        return "negative";
    } elsif ($n == 0) {
        return "zero";
    } else {
        return "positive";
    }
}

sub sum_array {
    my @nums = @_;
    my $total = 0;
    for my $n (@nums) {
        $total += $n;
    }
    return $total;
}

sub factorial {
    my ($n) = @_;
    return 1 if $n <= 1;
    my $result = 1;
    for my $i (2 .. $n) {
        $result *= $i;
    }
    return $result;
}

my $calc = Calculator->new(precision => 4);
print "add: ", $calc->add(3, 4), "\n";
print "classify(-5): ", classify(-5), "\n";
print "sum: ", sum_array(1, 2, 3, 4, 5), "\n";
print "factorial(5): ", factorial(5), "\n";
print "max: ", max(1, 7, 3), "\n";
